### PR TITLE
AEA-1148 Initial attempt at listing the gaps and tradeoffs present in v1

### PR DESCRIPTION
**Pull request by DavidMinarsch**
## Proposed changes

This PR adds to documentation the tradeoffs made in v1 as well as a list of ideas and rejected considerations.

By adding this to the docs we improve transparency and make it GitHub (issues) independent. We also implicitly force a more considered approach to feature design past v1. Eventually, this can be incorporated in a proper RFC process.

We (obviously) do *not* list issues relating framework internals or bugs here.

This PR also adds additional linter checks and resolves resulting issues.

This PR also fixes spelling in docstrings and resolves resulting issues.

## Fixes

It replaces the discussions in:
- #318
- #476
- #666
- #678
- #940
- #941
- #1033
- #1193
- #1319
- #1370
- #1643
- #1672
- #1708
- #1820
- #1979
- #2037
- #2096
- #2156
- #2170
- #2173
- #2219
- #2270

## Types of changes

What types of changes does your code introduce to agents-aea?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [ ] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [ ] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that code coverage does not decrease.
- [ ] I have checked that the documentation about the `aea cli` tool works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

-

### CODE REVIEW

File: docs/generic-storage.md [hunk 1]
```diff
@@ -22,7 +22,7 @@ One of the most useful cases is the integration of the dialogues subsystem and s
 ### Keep terminal state dialogues
 
 The Dialogues class has the optional boolean argument `keep_terminal_state_dialogues`
-which specifies whether a dialogue which has reached its terminal state is kept in memory or not. If `keep_terminal_state_dialogues` is `False`, dialogues that reach a terminal state are removed from memory and can not be used any more. If `keep_terminal_state_dialogues` is `True`, dialogues that reach a terminal state are kept.
+which specifies whether a dialogue which has reached its terminal state is kept in memory or not. If `keep_terminal_state_dialogues` is `False`, dialogues that reach a terminal state are removed from memory and can not be used any more. If `keep_terminal_state_dialogues` is `True`, dialogues that reach a terminal state are kept in memory or storage (if configured)E.
```
Comment (DavidMinarsch) [inline docs/generic-storage.md:25]: typo

Comment (solarw) [reply docs/generic-storage.md:25]: would be nice to say that it stored on agent stops and restored on agent start

**Review by solarw — APPROVED** [review]
LGTM

File: .pylintrc [hunk 2]
```diff
@@ -38,3 +36,10 @@ max-attributes=36
 
 [REFACTORING]
 max-nested-blocks=6
+
+[SPELLING]
+# uncomment to enable
```
Comment (DavidMinarsch) [inline .pylintrc:41]: Not currently enabled. Will do in a separate PR. All fixes are applied though.

File: docs/security.md [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+The following advice will help you when writing your own code:
+
+- Many potential common security vulnerabilities can be caught by static code analysis. We recommend you use `safety`, `pylint` and `bandit` to analyse your code.
+
+- Don't use relative import paths, these can lead to malicious code being executed.
+
+- Try to avoid using the `subprocess` module. If needed, make sure you sanitise commands passed to `subprocess`.
+
+- Try to avoid using the `pickle` module. Pickle should never be used for agent-to-agent communication protocols.
+
```
Comment (marcofavorito) [inline docs/security.md:13]: Some words on private key handling? Some thoughts (rewrite as you want):

By design, the framework prevents skill code from accessing private keys directly, as they are not reachable from the skill execution context through attribute getters. However, if the flag `-p` or the option `--password` are *not* used when generating or adding keys to an AEA project via the `aea` CLI tool, the private keys will be stored in _plaintext_. This clearly allows the skills to access them via interaction with the OS file system. We recommend to always specify a password to encrypt private keys by using the flag argument. (TODO) More details can be found in the `aea [command]` documentation.

File: docs/logging.md [hunk 4]
```diff
@@ -75,4 +75,66 @@ logging_config:
...
+    def stream():
+        print(dict(request.form))
+        return {}, 200
+
+    app.run()
+
+
+if __name__ == "__main__":
+    create_app()
+```
```
Comment (marcofavorito) [inline docs/logging.md:134]: https://github.com/fetchai/agents-aea/pull/2383

File: docs/limits.md [hunk 5]
```diff
@@ -0,0 +1,258 @@
+This document describes some of the limitations of `v1` of the AEA framework and tradeoffs made in its design.
+
+### Rejected ideas:
+
+#### Handlers implemented as behaviours:
+
+Handlers can be considered a special cases of a "behaviour that listens for specific events to happen".
+
+One could implement `Handler` classes in terms of `Behaviours`, after having implemented the feature that behaviours can be activated after an event happens (e.g. receiving a message of a certain protocol).
```
Comment (marcofavorito) [inline docs/limits.md:9]: maybe a sentence on why this has been rejected?

- better to have a clear separation of concerns, and avoid purely reactive (handlers) and proactive (behaviours) components are conflated into one concept.
- it adds complexity to behaviour development.
- other...

File: docs/limits.md [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+This approach is currently used and implemented across all the reference skills.
+
+Related, the topic of latency in transactions. State channels provide a solution. E.g. https://github.com/perun-network/go-perun. There could also be an interesting overlap with our protocols here.
+
+
+#### Unsolved problems in `Multiplexer` - `AgentLoop` interplay
+
+Problem 1: connection generates too many messages in a short amount of time, that are not consumed by the multiplexer
+Solution: Can be solved by slowing down connections receive method called, controlled by the inbox messages amount
+Side effects: Most of the connections should have an internal queue cause there is no synchronization between internal logic and multiplexer connection `receive` calls.
```
Comment (marcofavorito) [inline docs/limits.md:235]:
```suggestion
Side effects: Most of the connections should have an internal queue because there is no synchronization between internal logic and multiplexer connection `receive` calls.
```

File: docs/limits.md [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+
+
+Problem 2: the send method can take a long time (cause send retries logic in connection)
```
Comment (marcofavorito) [inline docs/limits.md:238]:
```suggestion
Problem 2: the send method can take a long time (because send retries logic in connection)
```

File: docs/limits.md [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+
+#### Containerized Agents
+
+Agent management, especially when many of them live on the same host, can be cumbersome. The framework should provide more utilities for these large-scale use cases. But a proper isolation of the agent environment is something that helps also simple use cases.
+
+A new software architecture, somehow inspired to the Docker system. The CLI only involves the initialization of the building of the agent (think of it as the specification of the `Dockerfile`: the `Agentfile`), but the actual build and run are done by the AEA engine, a daemon process analogous of the Docker Engine, which exposes APIs for these operations.
+
+Users and developers would potentially like to run many AEAs of different versions and with differences in the versions of their dependencies. It is not possible to import different versions of the same Python (PyPI) package in the same process in a clean way. However, in different processes this is trivial with virtual environments. It would be desirable to consider this in the context of a container solution for agents.
+
```
Comment (marcofavorito) [inline docs/limits.md:144]: maybe missing: flexible scaling of a single skill component depending on the workload

**Review by marcofavorito — APPROVED** [review]
LGTM, but first have a look at #2383 and ideally merge that PR before this PR.
